### PR TITLE
refactored using scala collection methods

### PR DESCRIPTION
**Pull request by rsalama**
Changed MailTFIDF.scala using scala collection methods. I was running on a vm, would need to remove the first(1000) on line 61

### CODE REVIEW

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 1]
```diff
@@ -1,11 +1,17 @@
-/*
+package org.apache.flink
+
+/**
  * Copyright 2015 data Artisans GmbH
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please do not change the license header. 
This code is not part of Apache Flink and does not belong to the ASF.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 2]
```diff
@@ -1,11 +1,17 @@
-/*
+package org.apache.flink
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please keep the original package

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 3]
```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
...
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+").filter(s => (isWord(s))))) first (1000) // TODO : remove first (1000) i was running on a vm
+
+    val cnt = mails.count
+    // val tfs = mails flatMap (m => m._2.groupBy(w => w).mapValues(_.size).map(e => (m._1, e._1, e._2)))
+    val tfs = mails flatMap (m => m._2.groupBy(w => w).map(e => (m._1, e._1, e._2.size)))
+    val dfs = (mails flatMap (m => m._2.toSet)) map (m => (m, 1)) groupBy(0) reduce ((l, r) => (l._1, l._2 + r._2))
+    val tfidfs = tfs.join(dfs).where(1).equalTo(0) { (l, r) => (l._1, l._2, l._3 * (cnt.toDouble / r._2) ) }
+
+    // mails.map(m => (m._1, m._2.mkString(" "))).writeAsText(s"file://${odir}/MAILS.csv", WriteMode.OVERWRITE)
+    tfs.writeAsText(s"file://${odir}/TFs.csv", WriteMode.OVERWRITE)
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please do not change the behavior of the job and only print the final result. 
This job is used as a reference implementation for the Flink training. We tried that all reference implementations follow a common pattern and I do not want to update the exercise descriptions. Thanks

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 4]
```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
...
+
+    val isWord = (s : String) => (!stopWords.contains(s) && wordPattern.findFirstIn(s).isDefined)
+
+    val mails = env.readCsvFile[(String, String)](
+      input,
+      lineDelimiter = "##//##",
+      fieldDelimiter = "#|#",
+      includedFields = Array(0,4)
+    // ) map (m => (m._1, m._2.toLowerCase.split("\\W+").collect{case s if isWord(s) => s})) first (1000)
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+").filter(s => (isWord(s))))) first (1000) // TODO : remove first (1000) i was running on a vm
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please resolve the TODO

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 5]
```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
...
+
+    val cnt = mails.count
+    // val tfs = mails flatMap (m => m._2.groupBy(w => w).mapValues(_.size).map(e => (m._1, e._1, e._2)))
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: No code in comments please.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 6]
```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
...
+    val wordPattern = "(\\p{Alpha})+".r
+
+    val isWord = (s : String) => (!stopWords.contains(s) && wordPattern.findFirstIn(s).isDefined)
+
+    val mails = env.readCsvFile[(String, String)](
+      input,
+      lineDelimiter = "##//##",
+      fieldDelimiter = "#|#",
+      includedFields = Array(0,4)
+    // ) map (m => (m._1, m._2.toLowerCase.split("\\W+").collect{case s if isWord(s) => s})) first (1000)
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: No code in comments please.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 7]
```diff
@@ -32,142 +30,48 @@ import scala.collection.mutable
...
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+").filter(s => (isWord(s))))) first (1000) // TODO : remove first (1000) i was running on a vm
+
+    val cnt = mails.count
+    // val tfs = mails flatMap (m => m._2.groupBy(w => w).mapValues(_.size).map(e => (m._1, e._1, e._2)))
+    val tfs = mails flatMap (m => m._2.groupBy(w => w).map(e => (m._1, e._1, e._2.size)))
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please add comments for the transformations (step-by-step if necessary).
This is a reference implementation should help people to learn Flink. 
So descriptive comments are more important than conciseness of code.

**Comment by fhueske** [comment]
Thanks for the PR @rsalama! I left a few comments in the diff view.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 8]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+      lineDelimiter = MBoxParser.MAIL_RECORD_DELIM,
+      fieldDelimiter = MBoxParser.MAIL_FIELD_DELIM,
+      includedFields = Array(0,4)
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+")  // convert message to lower case and split on word boundary
+      .filter(s => isWord(s)))) // retain only those strings that are words
+
+    // count mails in data set
+    val cnt = mails.count
+
+    // mails is a DataSet[String, Array[String]]
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Can you remove this comment and add the type to `val mails: DataSet[(String, Array[String)]`?

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 9]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+    // for each mail, use Array.groupBy(w => w) to create a Map[(String, Array[String])]
+    //    where the key is a string and the value an Array containing all strings.
+    //    Then use map on the entries of the Map to pick out:
+    //       e._1: the key and
+    //       e._2.size: the size of each Array[String] (number of times a string occurred)
+    val tfs = mails flatMap (m => m._2.groupBy(w => w).map(e => (m._1, e._1, e._2.length)))
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please define the types of the other DataSets as well.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 10]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+      fieldDelimiter = MBoxParser.MAIL_FIELD_DELIM,
+      includedFields = Array(0,4)
+    ) map (m => (m._1, m._2.toLowerCase.split("\\W+")  // convert message to lower case and split on word boundary
+      .filter(s => isWord(s)))) // retain only those strings that are words
+
+    // count mails in data set
+    val cnt = mails.count
+
+    // mails is a DataSet[String, Array[String]]
+    // for each mail, use Array.groupBy(w => w) to create a Map[(String, Array[String])]
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: should be `Map[String, Array[String]]` instead of `Map[(String, Array[String])]`

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 11]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+    //    where the key is a string and the value an Array containing all strings.
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: can you rephrase to: "where the key is a word and the value is an Array containing all occurrences of the word in the document."?

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 12]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+    //    Then use map on the entries of the Map to pick out:
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Can you put `groupBy` and `map` into separate lines and the comments next to each line?

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala (hunk 9 above)
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: Please add a comment what is computed here (frequency of words in a document). Will make it easier to follow the comments about how it is computed.

File: src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala [hunk 13]
```diff
@@ -37,137 +32,57 @@ import scala.collection.mutable
...
+    //       e._1: the key and
+    //       e._2.size: the size of each Array[String] (number of times a string occurred)
+    val tfs = mails flatMap (m => m._2.groupBy(w => w).map(e => (m._1, e._1, e._2.length)))
+
+    // Extract unique words by creating a set out of the Array[String] of each mail
+    // create a (m, 1) tuple for each string
+    val dfs = (mails flatMap (m => m._2.toSet)) // Extract unique words of each mail converting Array[String] to a Set[String]
+      .map (m => (m, 1))  // Create the (string, 1) initial record for each unique string
+      .groupBy(0).reduce ((l, r) => (l._1, l._2 + r._2)) // count the number of mails for each word
```
Comment (fhueske) [inline src/main/scala/com/dataartisans/flinktraining/exercises/dataset_scala/tf_idf/MailTFIDF.scala]: We can use `sum(1)` instead of `reduce ((l, r) => (l._1, l._2 + r._2))`

**Comment by fhueske** [comment]
Thanks @rsalama, the code looks good. Much better than the previous version.
I made a few minor style / comment suggestions.

Thanks, Fabian

**Comment by rsalama** [comment]
Fabian, 
incorporated your suggestions. Thanks for the suggestion on using sum(1)

**Comment by fhueske** [comment]
Hi Roberto, thanks for the update.
I was about to merge your PR, but then I noticed that the result has changed. 
I did a bit to debugging and more words seem to pass the filter. Can you have a look at what has changed? It is also possible that the current version is not working correctly.

Thanks, Fabian

**Comment by rsalama** [comment]
Hi Fabian, apologies for the errors. This version should fix that. The problems as you noted were:
1. needed to use WORD_PATTERN.unapplySeq(s) and not findFirstIn(s)
2. Changed split() to use \\s instead of \\W

Please let me know if this version is ok.
Thanks

PS, an alternative is to just use the pattern directly to do the match:

val isWord : String => Boolean = s => !STOP_WORDS.contains(s) && **WORD_PATTERN.pattern.matcher(s).matches**

**Comment by rsalama** [comment]
Hi Fabian, apologies for the errors. This version should fix that. The
problems as you noted were:
1. needed to use WORD_PATTERN.unapplySeq(s) and not findFirstIn(s)
2. Changed split() to use \s instead of \W

Please let me know if this version is ok.
Thanks

**Comment by rsalama** [comment]
Hi Fabian:
Did you have a chance to look at the changes I pushed last?
thanks

**Comment by fhueske** [comment]
Hi @rsalama, thanks for reminding me. 
Just merged your PR.
Thanks again for improving the exercises!

**Comment by rsalama** [comment]
cool, thanks!
